snr.py

```diff
@@ -0,0 +1,287 @@
+    npix : `~astropy.units.Quantity`, optional
+        Number of pixels under consideration for the signal with units of
+        pixels. Default is 1 * astropy.units.pixel.
+    n_background : `~astropy.units.Quantity`, optional
+        Number of pixels used in the background estimation with units of
+        pixels. Default is set to np.inf * astropy.units.pixel
+        such that there is no contribution of error due to background
+        estimation. This assumes that n_background will be >> npix.
+    background_rate : `~astropy.units.Quantity`, optional
+        Photons per pixel per second due to the backround/sky with units of
+        electrons/second/pixel.
+        Default is 0 * (astropy.units.electron /
+                        astropy.units.second / astropy.units.pixel)
+    darkcurrent_rate : `~astropy.units.Quantity`, optional
+        Electrons per pixel per second due to the dark current with units
+        of electrons/second/pixel.
+        Default is 0 * (astropy.units.electron /
+                        astropy.units.second / astropy.units.pixel)
+    readnoise : `~astropy.units.Quantity`, optional
+        Electrons per pixel from the read noise with units of electrons/pixel.
+        Default is 0 * astropy.units.electron / astropy.units.pixel.
+    gain : `~astropy.units.Quantity`, optional
+        Gain of the CCD with units of electrons/ADU. Default is
+        1 * astropy.units.electron / astropy.units.adu such that the
+        contribution to the error due to the gain is assumed to be small.
+    ad_err : `~astropy.units.Quantity`, optional
+        An estimate of the 1 sigma error within the A/D converter with units of
+        adu/pixel. Default is set to
+        sqrt(0.289) * astropy.units.adu / astropy.units.pixel
+        (Merline & Howell, 1995).
+
+    Returns
+    -------
+    t : `~astropy.units.Quantity`
+        The exposure time needed (in seconds) to achieve the given signal
+        to noise ratio.
+    """
+    readnoise = _get_shotnoise(readnoise)
+    gain_err = _get_shotnoise(gain * ad_err)
+
+    # solve t with the quadratic equation (pg. 57 of Howell 2000)
+    A = countrate ** 2
+    B = (-1) * snr ** 2 * (countrate + npix * (background_rate +
+                                               darkcurrent_rate))
+    C = (-1) * snr ** 2 * npix * readnoise ** 2
+
+    t = (-B + np.sqrt(B ** 2 - 4 * A * C)) / (2 * A)
+
+    if gain_err.value > 1 or not np.isfinite(n_background.value):
+        # solve t numerically
+        t = fsolve(_t_with_small_errs, t, args=(background_rate,
+                                                darkcurrent_rate,
+                                                gain_err, readnoise, countrate,
+                                                npix, n_background))
+        if not hasattr(t, 'unit'):
+            t = t * u.s
+
+    return t
+
+
+def test_howell_snr_calc():
+    """
+    A test to check that the math in howell_snr() is done correctly.
```

Comment:
Sphinx directive to xref a local function:
```suggestion
    A test to check that the math in :func:`howell_snr` is done correctly.
```